database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -14,57 +14,60 @@
 
 package com.firebase.ui.database;
 
+import android.support.annotation.NonNull;
+import android.support.v7.widget.RecyclerView;
 import android.util.Log;
 
 import com.google.firebase.database.DataSnapshot;
 import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
 import com.google.firebase.database.Query;
 import com.google.firebase.database.ValueEventListener;
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
-import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
+import java.util.ListIterator;
 import java.util.Map;
-import java.util.Set;
 
-class FirebaseIndexArray extends FirebaseArray {
-    private static final String TAG = FirebaseIndexArray.class.getSimpleName();
+public class FirebaseIndexArray extends FirebaseArray {
+    private static final String TAG = "FirebaseIndexArray";
 
-    private Query mQuery;
-    private ChangeEventListener mListener;
+    private DatabaseReference mDataRef;
     private Map<Query, ValueEventListener> mRefs = new HashMap<>();
     private List<DataSnapshot> mDataSnapshots = new ArrayList<>();
 
-    public FirebaseIndexArray(Query keyRef, Query dataRef) {
-        super(keyRef);
-        mQuery = dataRef;
+    /**
+     * @param keyQuery The Firebase location containing the list of keys to be found in {@code
+     *                 dataRef}. Can also be a slice of a location, using some combination of {@code
+     *                 limit()}, {@code startAt()}, and {@code endAt()}.
+     * @param dataRef  The Firebase location to watch for data changes. Each key key found at {@code
+     *                 keyQuery}'s location represents a list item in the {@link RecyclerView}.
+     */
+    public FirebaseIndexArray(Query keyQuery, DatabaseReference dataRef) {
+        super(keyQuery);
+        mDataRef = dataRef;
     }
 
     @Override
-    public void cleanup() {
-        super.cleanup();
-        Set<Query> refs = new HashSet<>(mRefs.keySet());
-        for (Query ref : refs) {
-            ref.removeEventListener(mRefs.remove(ref));
+    public void removeChangeEventListener(@NonNull ChangeEventListener listener) {
+        super.removeChangeEventListener(listener);
+        if (!isListening()) {
```

Comment:
For this method I would lock the entire body (removing the event listeners and clearing data/refs) and use the `mListeners` list as the lock.

It's also worth noting that `isListening()` itself is implemented as a check on mListeners, so that should be inside the lock.